--- /home/cris/projects/FreeBSD-src/usr.bin/truncate/truncate.c	2026-08-01 16:31:49.438539292 -0500
+++ truncate.c	2026-08-01 16:31:16.945702588 -0500
@@ -39,10 +39,10 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <stdio.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <unistd.h>
-
-#include <libutil.h>
+#include <util.h>
 
 static void	usage(void);
 
@@ -51,10 +51,9 @@
 {
 	struct stat sb;
 	mode_t omode;
-	off_t oflow, rsize, sz, tsize, round, off, len;
+	off_t oflow, rsize, sz, tsize, round;
 	uint64_t usz;
 	int ch, error, fd, oflags, r;
-	int do_dealloc;
 	int do_truncate;
 	int no_create;
 	int do_relative;
@@ -64,22 +63,17 @@
 	char *fname, *rname;
 
 	fd = -1;
-	rsize = tsize = sz = off = 0;
-	len = -1;
-	do_dealloc = no_create = do_relative = do_round = do_refer =
+	rsize = tsize = sz = 0;
+	no_create = do_relative = do_round = do_refer =
 	    got_size = 0;
 	do_truncate = 1;
 	error = r = 0;
 	rname = NULL;
-	while ((ch = getopt(argc, argv, "cdr:s:o:l:")) != -1)
+	while ((ch = getopt(argc, argv, "cr:s:")) != -1)
 		switch (ch) {
 		case 'c':
 			no_create = 1;
 			break;
-		case 'd':
-			do_dealloc = 1;
-			do_truncate = 0;
-			break;
 		case 'r':
 			do_refer = 1;
 			rname = optarg;
@@ -100,22 +94,6 @@
 				-(off_t)usz : (off_t)usz;
 			got_size = 1;
 			break;
-		case 'o':
-			if (expand_number(optarg, &usz) == -1 ||
-			    (off_t)usz < 0)
-				errx(EXIT_FAILURE,
-				    "invalid offset argument `%s'", optarg);
-
-			off = usz;
-			break;
-		case 'l':
-			if (expand_number(optarg, &usz) == -1 ||
-			    (off_t)usz <= 0)
-				errx(EXIT_FAILURE,
-				    "invalid length argument `%s'", optarg);
-
-			len = usz;
-			break;
 		default:
 			usage();
 			/* NOTREACHED */
@@ -131,8 +109,7 @@
 	 * the length argument must be set.  See usage() for allowed
 	 * invocations.
 	 */
-	if (argc < 1 || do_refer + got_size + do_dealloc != 1 ||
-	    (do_dealloc == 1 && len == -1))
+	if (argc < 1 || do_refer + got_size != 1)
 		usage();
 	if (do_refer == 1) {
 		if (stat(rname, &sb) == -1)
@@ -140,7 +117,7 @@
 		tsize = sb.st_size;
 	} else if (do_relative == 1 || do_round == 1)
 		rsize = sz;
-	else if (do_dealloc == 0)
+	else
 		tsize = sz;
 
 	if (no_create)
@@ -196,13 +173,6 @@
 		if (tsize < 0)
 			tsize = 0;
 
-		if (do_dealloc == 1) {
-			struct spacectl_range sr;
-
-			sr.r_offset = off;
-			sr.r_len = len;
-			r = fspacectl(fd, SPACECTL_DEALLOC, &sr, 0, &sr);
-		}
 		if (do_truncate == 1)
 			r = ftruncate(fd, tsize);
 		if (r == -1) {
@@ -219,9 +189,8 @@
 static void
 usage(void)
 {
-	fprintf(stderr, "%s\n%s\n%s\n",
+	fprintf(stderr, "%s\n%s\n",
 	    "usage: truncate [-c] -s [+|-|%|/]size[K|k|M|m|G|g|T|t] file ...",
 	    "       truncate [-c] -r rfile file ...",
-	    "       truncate [-c] -d [-o offset[K|k|M|m|G|g|T|t]] -l length[K|k|M|m|G|g|T|t] file ...");
 	exit(EXIT_FAILURE);
 }
